src/charm.py

```diff
@@ -2,202 +2,124 @@
 
 import logging
 from string import Template
-
+from charms.observability_libs.v1.kubernetes_service_patch import KubernetesServicePatch
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, MaintenanceStatus
+from ops.model import ActiveStatus, WaitingStatus, ModelError
+from ops.pebble import ServiceStatus
+from lightkube.models.core_v1 import ServicePort
 
-from oci_image import OCIImageResource, OCIImageResourceError
+logger = logging.getLogger(__name__)
 
 
 class CoreDNSCharm(CharmBase):
+    """CoreDNS Sidecar Charm"""
+
+    _COREDNS_CONTAINER = "coredns"
+
     def __init__(self, *args):
         super().__init__(*args)
+
         if not self.unit.is_leader():
             # We can't do anything useful when not the leader, so do nothing.
-            self.model.unit.status = ActiveStatus()
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
             return
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, 'coredns-image')
-        for event in [self.on.install,
-                      self.on.leader_elected,
-                      self.on.upgrade_charm,
-                      self.on.config_changed]:
-            self.framework.observe(event, self.main)
-        self.framework.observe(self.on.dns_provider_relation_joined, self.provide_dns)
-
-    def main(self, event):
-        try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            self.model.unit.status = e.status
+
+        dns_udp = ServicePort(53, protocol="UDP", name="dns")
+        dns_tcp = ServicePort(53, protocol="TCP", name="dns-tcp")
+        metrics = ServicePort(9153, protocol="TCP", name="metrics")
+        self.service_patcher = KubernetesServicePatch(self, [dns_udp, dns_tcp, metrics])
+
+        self.framework.observe(
+            self.on.coredns_pebble_ready, self._on_coredns_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on.dns_provider_relation_changed,
```

Comment:
good point. Tying it to the created hook makes sense. Im trying to figure out how to make my unit tests pass with that as the harness does not seem to have the implicit ingress adress present in the data. With relation-changed, I could modify the relation data manually and inject the ingress address before triggering the hook.

Reply:
@johnsca I have added code to config changed that republishes relation data so that the domain is updated, and added a unit test for that case. I also changed the relation data to use app level data instead of unit level data.